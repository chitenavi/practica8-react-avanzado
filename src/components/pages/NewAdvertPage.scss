@import '../../scss/variables';

.formNewAd {
  display: flex;
  flex-direction: column;
  margin: 0 auto;
  width: 400px;
  gap: 1rem;
  background-color: $bg-color-ter;
  box-shadow: $shadow-primary;
  border-radius: 1rem;
  padding: 1rem;

  &-group {
    display: flex;
    align-items: center;
    justify-content: space-between;
    text-align: center;
  }

  &-field {
    &--label {
      font-size: 1.1rem;
      margin-right: 0.2rem;
    }

    &.centered {
      text-align: center;
    }

    &--img {
      margin: 0 auto;
      height: 150px;
    }
  }
  &-error {
    margin-top: 1rem;
    text-align: center;
  }
}

.newAdPage {
  &-error {
    margin-top: 1rem;
    text-align: center;
    max-width: 400px;
    margin: 2rem auto;
  }
}
