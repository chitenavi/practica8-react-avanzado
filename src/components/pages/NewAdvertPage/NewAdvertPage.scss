@import '../../../scss/variables';

.newAdPage .form {
  display: flex;
  flex-direction: column;
  margin: 0 auto;
  width: 400px;
  gap: 1rem;
  background-color: $bg-color-ter;
  box-shadow: $shadow-primary;
  border-radius: 1rem;
  padding: 1rem;

  // &-group {
  //   display: flex;
  //   align-items: center;
  //   justify-content: space-between;
  //   text-align: center;
  // }

  & .form-field {
    &--label {
      font-size: 1.1rem;
      margin-right: 0.2rem;
    }

    &--img {
      margin: 0 auto;
      height: 150px;
    }
  }
  &-error {
    margin-top: 1rem;
    text-align: center;
  }

  &-button {
    margin: 0 auto;
  }
}

.newAdPage {
  &-error {
    margin-top: 1rem;
    text-align: center;
    max-width: 400px;
    margin: 2rem auto;
  }
}
