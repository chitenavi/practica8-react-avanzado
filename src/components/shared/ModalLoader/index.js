import ModalLoader from './ModalLoader';

export default ModalLoader;
