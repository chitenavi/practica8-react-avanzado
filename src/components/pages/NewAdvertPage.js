import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import { Radio, Input, InputNumber, Alert } from 'antd';
import useForm from '../../hooks/useForm';
import MainLayout from '../layout/MainLayout';
import SelectTags from '../shared/SelectTags';
import Button from '../shared/Button';
import ModalLoader from '../shared/ModalLoader';
import FileImageLoad from '../shared/FileImageLoad';
import { createAdvert } from '../../api/adverts';
import './NewAdvertPage.scss';

const NewAdvertPage = () => {
  const history = useHistory();
  const [selectedFile, setSelectedFile] = useState(null);
  const [error, setError] = useState(null);
  const [submitting, setSubmitting] = useState(false);
  const [form, onChange] = useForm({
    name: '',
    sale: true,
    tags: [],
    price: 0,
  });
  const { name, sale, tags, price } = form;

  const canSubmit = () => {
    return name && tags.length && price;
  };

  const onSubmitForm = async e => {
    e.preventDefault();
    setSubmitting(true);
    setError(null);
    try {
      const formData = new FormData();
      Object.keys(form).forEach(key => {
        if (key === 'tags') tags.forEach(val => formData.append(key, val));
        else formData.append(key, form[key]);
      });

      if (selectedFile) formData.append('photo', selectedFile);

      const newAd = await createAdvert(formData);
      setSubmitting(false);
      history.push(`/advert/${newAd._id}`);
    } catch (err) {
      setSubmitting(false);
      setError(err);
    }
  };

  // TODO: Create image thumbnail in the app when user select one
  return (
    <MainLayout title="New Advert">
      <form
        method="POST"
        encType="multipart/form-data"
        onSubmit={onSubmitForm}
        className="formNewAd"
      >
        <div className="formNewAd-field">
          <Input
            onChange={onChange}
            name="name"
            value={name}
            placeholder="Advert name"
          />
        </div>
        <div className="formNewAd-group">
          <div className="formNewAd-field">
            <span className="formNewAd-field--label">Price: </span>
            <InputNumber
              name="price"
              onChange={value => {
                onChange({ target: { value, name: 'price' } });
              }}
              min={0}
              max={10000}
              value={price}
            />
          </div>
          <div className="formNewAd-field">
            <span className="formNewAd-field--label">Type: </span>
            <Radio.Group name="sale" onChange={onChange} value={sale}>
              <Radio style={{ color: 'white' }} value>
                Sale
              </Radio>
              <Radio style={{ color: 'white' }} value={false}>
                Buy
              </Radio>
            </Radio.Group>
          </div>
        </div>
        <div className="formNewAd-field centered">
          <SelectTags onChange={onChange} defaultTags={tags} />
        </div>
        <div className="formNewAd-field centered">
          <FileImageLoad
            label="Select a single image file"
            onFileSelect={file => {
              // console.log(URL.createObjectURL(file));
              setSelectedFile(file);
            }}
          />
        </div>

        <div className="formNewAd-field centered">
          <Button type="submit" className="secondary" disabled={!canSubmit()}>
            Create Advert
          </Button>
        </div>
      </form>
      {submitting && <ModalLoader />}
      {error && (
        <div className="newAdPage-error">
          <Alert message={error.message} type="error" />
        </div>
      )}
    </MainLayout>
  );
};

export default NewAdvertPage;
