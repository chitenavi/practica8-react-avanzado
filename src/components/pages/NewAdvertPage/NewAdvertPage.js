import React from 'react';
import PropTypes from 'prop-types';
import Form, {
  InputCustom,
  SelectCustom,
  InputImage,
  RadioGroup,
} from '../../shared/Form';
import MainLayout from '../../layout/MainLayout';
import ModalLoader from '../../shared/ModalLoader';
import ErrorMessage from '../../errors/ErrorMessage';
import { formNewAd } from '../../../config';

import './NewAdvertPage.scss';

const NewAdvertPage = ({ onCreate, error, loading }) => {
  const onSubmitForm = data => {
    const formData = new FormData();
    Object.keys(data).forEach(key => {
      if (key !== 'file' && key !== 'validateFields') {
        if (key === 'tags') data[key].forEach(val => formData.append(key, val));
        else formData.append(key, data[key]);
      }
    });

    if (data.file) formData.append('image', data.file);

    onCreate(formData);
  };

  return (
    <MainLayout title="New Advert">
      <div className="newAdPage">
        {loading && <ModalLoader />}
        <Form
          method="POST"
          encType="multipart/form-data"
          onSubmit={onSubmitForm}
          initialValue={formNewAd}
          submitLabel="Create Ad"
        >
          <InputCustom type="text" name="name" placeholder="Advert name" />
          <div className="form-field--group">
            <InputCustom type="price" name="price" />
            <RadioGroup label="Type" name="sale" options={['sale', 'buy']} />
          </div>
          <SelectCustom placeholder="Select Tags" name="tags" />
          <InputImage name="file" className="centered" />
        </Form>
        {error && (
          <ErrorMessage message={error.message} className="newAdPage-error" />
        )}
      </div>
    </MainLayout>
  );
};

NewAdvertPage.propTypes = {
  loading: PropTypes.bool.isRequired,
  error: PropTypes.objectOf(PropTypes.any),
  onCreate: PropTypes.func.isRequired,
};

NewAdvertPage.defaultProps = {
  error: null,
};

export default NewAdvertPage;
