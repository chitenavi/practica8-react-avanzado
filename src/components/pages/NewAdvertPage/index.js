import { connect } from 'react-redux';
import NewAdvertPage from './NewAdvertPage';

import { getUi } from '../../../store/selectors';
import { createAdvert } from '../../../store/actions';

const mapStateToProps = getUi;
const mapDispatchToProps = dispatch => ({
  onCreate: advertData => dispatch(createAdvert(advertData)),
});

export default connect(mapStateToProps, mapDispatchToProps)(NewAdvertPage);
