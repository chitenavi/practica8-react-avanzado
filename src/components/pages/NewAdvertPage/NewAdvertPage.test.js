import React from 'react';
import { shallow } from 'enzyme';

import NewAdvertPage from './NewAdvertPage';
import ModalLoader from '../../shared/ModalLoader';

describe('NewAdvertPage', () => {
  const onCreate = jest.fn();
  const loading = true;
  const error = null;

  const render = () =>
    shallow(
      <NewAdvertPage onCreate={onCreate} loading={loading} error={error} />,
    );

  const wrapper = render();

  it('should render a loader while loading prop is true', () => {
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.contains(<ModalLoader />)).toBe(true);
  });

  it('should render page with error', () => {
    wrapper.setProps({ loading: false, error: { message: 'error' } });

    expect(wrapper.exists()).toBe(true);
    expect(wrapper.exists('.newAdPage-error')).toBe(true);
  });

  it('snapshot with error', () => {
    expect(wrapper.exists('.newAdPage-error')).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render new advert page', () => {
    wrapper.setProps({ error: null });
    expect(wrapper.exists()).toBe(true);
  });

  it('should call onCreate when submit form', () => {
    const data = {};

    const form = wrapper.find('Form');
    form.simulate('submit', data);

    expect(wrapper.exists()).toBe(true);
    expect(onCreate).toHaveBeenCalled();
  });

  it('snapshot before advert creation', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
