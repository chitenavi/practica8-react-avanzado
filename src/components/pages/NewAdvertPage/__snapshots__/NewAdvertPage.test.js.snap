// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewAdvertPage snapshot before advert creation 1`] = `
<MainLayout
  title="New Advert"
>
  <div
    className="newAdPage"
  >
    <Form
      encType="multipart/form-data"
      initialValue={
        Object {
          "file": null,
          "name": "",
          "price": 0,
          "sale": true,
          "tags": Array [],
          "validateFields": Array [
            "name",
            "tags",
            "price",
          ],
        }
      }
      method="POST"
      onSubmit={[Function]}
      submitLabel="Create Ad"
    >
      <InputCustom
        className=""
        name="name"
        placeholder="Advert name"
        type="text"
      />
      <div
        className="form-field--group"
      >
        <InputCustom
          className=""
          name="price"
          placeholder=""
          type="price"
        />
        <RadioGroup
          className=""
          label="Type"
          name="sale"
          options={
            Array [
              "sale",
              "buy",
            ]
          }
        />
      </div>
      <SelectCustom
        className=""
        name="tags"
        placeholder="Select Tags"
      />
      <SelectCustom
        className="centered"
        name="file"
      />
    </Form>
  </div>
</MainLayout>
`;

exports[`NewAdvertPage snapshot with error 1`] = `
<MainLayout
  title="New Advert"
>
  <div
    className="newAdPage"
  >
    <Form
      encType="multipart/form-data"
      initialValue={
        Object {
          "file": null,
          "name": "",
          "price": 0,
          "sale": true,
          "tags": Array [],
          "validateFields": Array [
            "name",
            "tags",
            "price",
          ],
        }
      }
      method="POST"
      onSubmit={[Function]}
      submitLabel="Create Ad"
    >
      <InputCustom
        className=""
        name="name"
        placeholder="Advert name"
        type="text"
      />
      <div
        className="form-field--group"
      >
        <InputCustom
          className=""
          name="price"
          placeholder=""
          type="price"
        />
        <RadioGroup
          className=""
          label="Type"
          name="sale"
          options={
            Array [
              "sale",
              "buy",
            ]
          }
        />
      </div>
      <SelectCustom
        className=""
        name="tags"
        placeholder="Select Tags"
      />
      <SelectCustom
        className="centered"
        name="file"
      />
    </Form>
    <ErrorMessage
      className="newAdPage-error"
      message="error"
    />
  </div>
</MainLayout>
`;
